File: client/test/stream_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:friday_client/friday/friday.dart';

import 'support.dart';

void main() {
  test('yields the updates then the answer, and ends', () async {
    final bytes = FakeByteSource(
      chunks: [
        sse('update', seq: 1, text: 'Let me take a look.'),
        sse('update', seq: 2, text: 'Still working on it.'),
        sse('final', seq: 3, text: 'Two of them look risky.'),
      ],
    );
    final server = FakeServer();
    final api = loggedIn(server, bytes: bytes);
    await authorise(api);

    final got = await api.streamChat('chat_01DDD').toList();

    expect(got.map((e) => e.text), [
      'Let me take a look.',
      'Still working on it.',
      'Two of them look risky.',
    ]);
    expect(got.last.isTerminal, isTrue);
  });

  // The stream is the one endpoint the browser's EventSource cannot reach,
  // because it cannot set this header. It must be sent.
  test('presents the token as a header, never in the URL', () async {
    final bytes = FakeByteSource(chunks: [sse('final', seq: 1, text: 'done')]);
    final server = FakeServer();
    final api = loggedIn(server, bytes: bytes);
    await authorise(api);

    await api.streamChat('chat_01DDD').toList();

    final opened = bytes.opened.single;
    expect(opened.headers['Authorization'], 'Bearer a-token');
    expect(opened.headers['Accept'], 'text/event-stream');
    expect(opened.url.toString(), isNot(contains('a-token')));
    expect(opened.url.path, '/v1/chats/chat_01DDD/stream');
  });

  // Hearing the same thing twice would mean a voice client repeating itself.
  test('asks to resume from where it left off', () async {
    final bytes = FakeByteSource(chunks: [sse('final', seq: 4, text: 'rest')]);
    final server = FakeServer();
    final api = loggedIn(server, bytes: bytes);
    await authorise(api);

    await api.streamChat('chat_01DDD', resumeFrom: 3).toList();

    expect(bytes.opened.single.headers['Last-Event-ID'], '3');
  });

  test('does not ask to resume from the beginning', () async {
    final bytes = FakeByteSource(chunks: [sse('final', seq: 1, text: 'done')]);
    final server = FakeServer();
    final api = loggedIn(server, bytes: bytes);
    await authorise(api);

    await api.streamChat('chat_01DDD').toList();

    expect(bytes.opened.single.headers.containsKey('Last-Event-ID'), isFalse);
  });

  test('refuses to open a stream without a token', () async {
    final server = FakeServer();
    final api = FridayApi(
      baseUrl: Uri.parse('https://friday.test'),
      httpClient: server.client,
      byteSource: FakeByteSource(),
    );

    await expectLater(
      api.streamChat('chat_01DDD'),
      emitsError(isA<NotAuthenticated>()),
    );
  });

  // A refusal arrives as a status and a short JSON body rather than as
  // events, so it has to be read and turned into the same exception the
  // ordinary calls produce.
  test('turns a refused stream into the matching failure', () async {
    final cases = {
      404: isA<NotFound>(),
      401: isA<NotAuthenticated>(),
      500: isA<ServerFailure>(),
    };

    for (final entry in cases.entries) {
      final bytes = FakeByteSource(
        status: entry.key,
        chunks: ['{"error":"No such chat."}'],
      );
      final server = FakeServer();
      final api = loggedIn(server, bytes: bytes);
      await authorise(api);

      await expectLater(
        api.streamChat('chat_01ZZZ'),
        emitsError(entry.value),
        reason: 'status ${entry.key}',
      );
    }
  });

  test('reports a connection that could not be opened', () async {
    final bytes = FakeByteSource(error: const SocketishError());
    final server = FakeServer();
    final api = loggedIn(server, bytes: bytes);
    await authorise(api);

    await expectLater(
      api.streamChat('chat_01DDD'),
      emitsError(isA<Unreachable>()),
    );
  });

  // A chat that finished before anyone listened replays what it said and
  // how it ended, rather than producing an empty stream.
  test('replays a chat that had already finished', () async {
    final bytes = FakeByteSource(
      chunks: [
        sse('update', seq: 1, text: 'one'),
        sse('update', seq: 2, text: 'two'),
        sse('final', seq: 3, text: 'the answer'),
      ],
    );
    final server = FakeServer();
    final api = loggedIn(server, bytes: bytes);
    await authorise(api);

    final got = await api.streamChat('chat_01DDD').toList();
    expect(got.length, 3);
    expect(got.last.kind, EventKind.finalAnswer);
  });

  test('ends on a failure with the reason, which is what is spoken', () async {
    final bytes = FakeByteSource(
      chunks: [
        sse('update', seq: 1, text: 'trying'),
        sse('error', text: 'I could not reach GitLab.'),
      ],
    );
    final server = FakeServer();
    final api = loggedIn(server, bytes: bytes);
    await authorise(api);

    final got = await api.streamChat('chat_01DDD').toList();

    expect(got.last.kind, EventKind.error);
    expect(got.last.text, 'I could not reach GitLab.');
    expect(got.last.isTerminal, isTrue);
  });

  test('ends on cancellation, which is what "stop" produces', () async {
    final bytes = FakeByteSource(
      chunks: [
        sse('update', seq: 1, text: 'working'),
        sse('cancelled'),
      ],
    );
    final server = FakeServer();
    final api = loggedIn(server, bytes: bytes);
    await authorise(api);

    final got = await api.streamChat('chat_01DDD').toList();
    expect(got.last.kind, EventKind.cancelled);
    expect(got.last.isTerminal, isTrue);
  });

  test('closing releases the stream source', () async {
    final bytes = FakeByteSource();
    final server = FakeServer();
    final api = loggedIn(server, bytes: bytes);

    api.close();
    expect(bytes.closed, isTrue);
  });
}

/// SocketishError : Stands in for whatever the transport throws.
class SocketishError implements Exception {
  const SocketishError();
}

File: internal/api/chats/stream_test.go
package chats_test

import (
	"bufio"
	"context"
	"encoding/json"
	"net/http"
	"strings"
	"testing"
	"time"

	"github.com/DhanushRamesh/personal-assistant/internal/api/apitest"
	"github.com/DhanushRamesh/personal-assistant/internal/api/chats"
	"github.com/DhanushRamesh/personal-assistant/internal/chat"
	"github.com/DhanushRamesh/personal-assistant/internal/provider"
)

// sseEvent : One parsed server-sent event.
type sseEvent struct {
	ID    string
	Event string
	Data  chats.Event
}

// readStream : Reads events from a live SSE response until it closes.
func readStream(t *testing.T, body *bufio.Reader) []sseEvent {
	t.Helper()

	var got []sseEvent
	var current sseEvent
	var haveData bool

	for {
		line, err := body.ReadString('\n')
		if line != "" {
			line = strings.TrimRight(line, "\r\n")
			switch {
			case line == "":
				if haveData {
					got = append(got, current)
				}
				current, haveData = sseEvent{}, false
			case strings.HasPrefix(line, "id: "):
				current.ID = strings.TrimPrefix(line, "id: ")
			case strings.HasPrefix(line, "event: "):
				current.Event = strings.TrimPrefix(line, "event: ")
			case strings.HasPrefix(line, "data: "):
				payload := strings.TrimPrefix(line, "data: ")
				if err := json.Unmarshal([]byte(payload), &current.Data); err != nil {
					t.Fatalf("unmarshal %q: %v", payload, err)
				}
				haveData = true
			case strings.HasPrefix(line, ":"):
				// A heartbeat comment.
			}
		}
		if err != nil {
			return got
		}
	}
}

// streamChat : Opens a stream against a live test server and reads it to the
// end.
func streamChat(t *testing.T, e *apitest.Env, base, id string, headers map[string]string) []sseEvent {
	t.Helper()

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, base+"/v1/chats/"+id+"/stream", nil)
	if err != nil {
		t.Fatalf("request: %v", err)
	}
	req.Header.Set("Authorization", "Bearer "+e.Token)
	for k, v := range headers {
		req.Header.Set(k, v)
	}

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		t.Fatalf("stream: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		t.Fatalf("stream status = %d, want 200", resp.StatusCode)
	}
	if ct := resp.Header.Get("Content-Type"); ct != "text/event-stream" {
		t.Errorf("Content-Type = %q, want text/event-stream", ct)
	}
	return readStream(t, bufio.NewReader(resp.Body))
}

// live : An environment served on a real listener, which SSE needs in order
// to flush a response progressively.
func live(t *testing.T, p provider.Provider) (*apitest.Env, string) {
	t.Helper()
	e := apitest.NewWith(t, apitest.Options{Provider: p})
	return e, e.Live(t)
}

// A client listening while a chat runs hears each message as it is produced,
// then the answer, and the stream ends.
func TestStreamDeliversMessagesThenTheAnswer(t *testing.T) {
	updates := []string{"Let me take a look.", "Still working on it."}
	e, base := live(t, &provider.Stub{Updates: updates, Delay: 40 * time.Millisecond})

	created := e.CreateChat(t, "/v1/chats", "check my merge requests")
	got := streamChat(t, e, base, created.ID, nil)

	if len(got) < len(updates)+1 {
		t.Fatalf("got %d events, want %d updates and a final: %+v", len(got), len(updates), got)
	}

	for i, want := range updates {
		if got[i].Event != "update" {
			t.Errorf("event %d kind = %q, want update", i, got[i].Event)
		}
		if got[i].Data.Text != want {
			t.Errorf("event %d text = %q, want %q", i, got[i].Data.Text, want)
		}
		if got[i].ID == "" {
			t.Errorf("event %d has no id, so a client cannot resume from it", i)
		}
	}

	last := got[len(got)-1]
	if last.Event != "final" {
		t.Fatalf("last event = %q, want final", last.Event)
	}
	if !strings.Contains(last.Data.Text, "check my merge requests") {
		t.Errorf("final text = %q, want the answer", last.Data.Text)
	}
}

// A chat that finished before anyone connected must still deliver everything
// it said and how it ended, rather than an empty stream.
func TestStreamReplaysAFinishedChat(t *testing.T) {
	updates := []string{"one", "two"}
	e, base := live(t, &provider.Stub{Updates: updates})

	created := e.CreateChat(t, "/v1/chats?wait=5s", "already done")
	if created.Status != string(chat.StatusCompleted) {
		t.Fatalf("status = %q, want the chat finished before streaming", created.Status)
	}

	got := streamChat(t, e, base, created.ID, nil)

	if len(got) != len(updates)+1 {
		t.Fatalf("got %d events, want %d replayed updates and a final: %+v", len(got), len(updates), got)
	}
	if got[len(got)-1].Event != "final" {
		t.Errorf("last event = %q, want final", got[len(got)-1].Event)
	}
}

// A reconnecting client resumes rather than hearing everything twice, which
// for a voice client would mean repeating itself.
func TestStreamResumesFromLastEventID(t *testing.T) {
	updates := []string{"one", "two", "three"}
	e, base := live(t, &provider.Stub{Updates: updates})

	created := e.CreateChat(t, "/v1/chats?wait=5s", "resume me")

	got := streamChat(t, e, base, created.ID, map[string]string{"Last-Event-ID": "2"})

	for _, ev := range got {
		if ev.Data.Seq > 0 && ev.Data.Seq <= 2 {
			t.Errorf("event %d was replayed despite Last-Event-ID: 2", ev.Data.Seq)
		}
	}
	if len(got) == 0 || got[len(got)-1].Event != "final" {
		t.Errorf("the outcome was not delivered: %+v", got)
	}
}

// A failed chat ends the stream with the reason, which is what the user hears.
func TestStreamEndsWithTheFailure(t *testing.T) {
	const reason = "I could not reach GitLab."
	e, base := live(t, &provider.Stub{Updates: []string{"trying"}, FailWith: reason})

	created := e.CreateChat(t, "/v1/chats", "will fail")
	got := streamChat(t, e, base, created.ID, nil)

	last := got[len(got)-1]
	if last.Event != "error" {
		t.Fatalf("last event = %q, want error: %+v", last.Event, got)
	}
	if last.Data.Text != reason {
		t.Errorf("error text = %q, want %q", last.Data.Text, reason)
	}
}

// Cancelling ends the stream, which is what saying "stop" must do.
func TestStreamEndsWhenTheChatIsCancelled(t *testing.T) {
	e, base := live(t, &provider.Stub{
		Updates: []string{"a", "b", "c", "d", "e"},
		Delay:   60 * time.Millisecond,
	})

	created := e.CreateChat(t, "/v1/chats", "stop me midway")

	done := make(chan []sseEvent, 1)
	go func() { done <- streamChat(t, e, base, created.ID, nil) }()

	time.Sleep(120 * time.Millisecond)
	rec := e.Do(t, http.MethodPost, "/v1/chats/"+created.ID+"/cancel", "")
	if rec.Code != http.StatusAccepted {
		t.Fatalf("cancel: status %d: %s", rec.Code, rec.Body)
	}

	select {
	case got := <-done:
		if len(got) == 0 {
			t.Fatal("stream produced nothing")
		}
		if last := got[len(got)-1].Event; last != "cancelled" {
			t.Errorf("last event = %q, want cancelled: %+v", last, got)
		}
	case <-time.After(10 * time.Second):
		t.Fatal("stream did not close after the chat was cancelled")
	}
}

func TestStreamOfAnUnknownChatIsNotFound(t *testing.T) {
	e := apitest.New(t)

	for _, id := range []string{chat.NewID(), "nonsense"} {
		rec := e.Get(t, "/v1/chats/"+id+"/stream")
		if rec.Code != http.StatusNotFound {
			t.Errorf("stream %s: status = %d, want 404", id, rec.Code)
		}
	}
}

// Two listeners on one chat both hear everything, as a phone and a desktop
// might.
func TestTwoListenersBothHearEverything(t *testing.T) {
	e, base := live(t, &provider.Stub{Updates: []string{"one", "two"}, Delay: 50 * time.Millisecond})

	created := e.CreateChat(t, "/v1/chats", "heard twice")

	first := make(chan []sseEvent, 1)
	second := make(chan []sseEvent, 1)
	go func() { first <- streamChat(t, e, base, created.ID, nil) }()
	go func() { second <- streamChat(t, e, base, created.ID, nil) }()

	for i, ch := range []chan []sseEvent{first, second} {
		select {
		case got := <-ch:
			if len(got) == 0 || got[len(got)-1].Event != "final" {
				t.Errorf("listener %d did not receive the answer: %+v", i, got)
			}
		case <-time.After(10 * time.Second):
			t.Fatalf("listener %d never finished", i)
		}
	}
}
